### PR TITLE
Add wrap files for tclap-1.2.1

### PR DESCRIPTION
**Pull request by ManuelAtWork**
Here is my working

- wrapfile
- patch (3x meson.build files)

for TCLAP v1.2.1.

### CODE REVIEW

File: upstream.wrap [hunk 1]
```diff
@@ -0,0 +1,10 @@
+[wrap-file]
+directory = tclap-1.2.1
+source_url = https://sourceforge.net/projects/tclap/files/tclap-1.2.1.tar.gz
+source_filename = tclap-1.2.1.tar.gz
+source_hash = 9f9f0fe3719e8a89d79b6ca30cf2d16620fba3db5b9610f9b51dd2cd033deebb
+
+#TODO: Add patch when available in Meson's WrapDB.
```
Comment (jpakkane) [inline upstream.wrap]: You don't need to have this here. It is in fact counterproductive. Wrap tooling will add these automatically.

Comment (ManuelAtWork) [reply upstream.wrap]: Do you mean only the comment? Or the ``source_filename`` and ``source_hash`` too?

Comment (jpakkane) [reply upstream.wrap]: The comment.

Comment (ManuelAtWork) [reply upstream.wrap]: Just amended&pushed my commit because I forgot to include a cpp file. (Sorry ...)
Removed the comment as suggested.

**Comment by jpakkane** [comment]
The test file is nice but we have a rule that wrapdb files must not contain any additional sources. There are a bunch of other rules as well, and the easiest way is to use the [mesonwrap](https://github.com/mesonbuild/mesonwrap/). Once you have it installed it's a matter of running this command:

    mesonwrap review --pull-request 1 tclap

**Comment by ManuelAtWork** [comment]
True, but there also is a rule that unit tests of the project shall be wrapped. However, the projcet's tests rely on shell scripts and I have not found a way to encapsulate these in ``executable``s, so that we could use them as a ``test``. This leaves us in the following dilemma:

- Break the no-code rule.
- Break the wrap-unit-test guideline.
- Find a way for Meson to encapsulate external commands as an ``executable``.
- Declare TCLAP as non-wrappable.

The supplied ``meson_build_wrapper.cpp`` is self-containing and does not interfere with the project's code in any way. So although it _technically_ is an additional source, it is _conceptually_ part of the build configuration.  (And therefore not covered by the ratio legis of the rule you mentioned.)

So how do we proceed? From my perspective, any solution would be ok.

Regardless of what we do here, I am planning to  contribute my test wrapper to the TCLAP project to be included in future versions. (And suggest to switch from Autotools to Meson.)

``mesonwrap`` appeared to me as a tool for the WrapDB maintainers, which I am not. But if you say it is helpful, I will look into it. :-)

**Comment by jpakkane** [comment]
You can use the output of `find_program` in tests. So something like `prog = find_program('script.sh')` and then `test('name', prog, ...)`. However, that is not portable to Windows. In this case it is fine to skip the unit tests altogether.

**Comment by ManuelAtWork** [comment]
I have amended the commit and implemented the tests using ```find_program``. Some checks disable tests when necessary prerequisites are not available (tested on MSVC 2019).

Do I need to open a new ticket when creating a wrap for the next version (1.2.2) of TCLAP? Or do I simply create a new branch and issue a pull request with a comment?

**Comment by jpakkane** [comment]
> I have amended the commit and implemented the tests using ```find_program``.

Looks good. However the code is indented with tabs, which is prohibited. I again highly recommend using the `mesonwrap` tool mentioned above. It will allow you to get all of these minor issues fixed without needing a human being in the loop. Basically what I do when reviewing is first run the tool and only after it gives the green light do I start looking at the actual content of the files.

> Do I need to open a new ticket when creating a wrap for the next version (1.2.2) of TCLAP? Or do I simply create a new branch and issue a pull request with a comment?

There is no need to create bugs for existing repos. Just do:

- branch `master`
- add your changes
- create a pull request against `master`
- the wrap reviewer will create the branch and retarget the MR

**Comment by ManuelAtWork** [comment]
Thank you for the quick reply. I tried to get ``mesonwrap`` to work, but was unsuccessful. It is not included in my distro, and my Python skills are not good to enough to get it to work. (Maybe you could include an install.txt or a 5-minute tutorial in the mesonwrap project?)

**Comment by jpakkane** [comment]
Install Python's virtualenv, go to the mesonwrap source root and then:

```
virtualenv venv
bash
source venv/bin/activate
python3 setup.py install
# install dependencies with pip, repeat, pip will install to the virtualenv Python, not the system one
mesonwrap review --pull-request 1 tclap
# repeat until it passes
exit
```

File: tests/meson.build [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+endif
+
+
+# The test scripts need $srcdir pointing to the directory of the 'test*.out' files.
+srcdir_param = 'srcdir='+meson.current_source_dir()
+
+
+tests = [
+	# <Name>     <test script>  <executable> 
+	[ 'test 1',  './test1.sh',  test1  ],
```
Comment (eli-schwartz) [inline tests/meson.build]: No need for the `./` here, the final lookup looks like this:

```
Program ./test1.sh found: YES (/tmp/tmpi08csggu/src/tclap-1.2.1/./tests/./test1.sh)
```

meson already uses the full path.

File: meson.build [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+tclap_dep = declare_dependency(
+	version : '1.2.1',
+	include_directories : 'include')
+
+# --
+# Examples
+# --
+
+subdir('./examples')
```
Comment (eli-schwartz) [inline meson.build]: No need for the `./` here either.

File: tests/meson.build [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+
+foreach t : tests
+	test(t[0], find_program(t[1]), env : srcdir_param, depends : t[2], is_parallel : false)
+endforeach
+
+
+
+
+
+
```
Comment (eli-schwartz) [inline tests/meson.build]: Empty lines here look ugly?

**Review by eli-schwartz — CHANGES_REQUESTED** [review]
Also all the tests fail. *all of them*

Looking at the script runners and the contents of tmp.out directly, it says

```
/tmp/tclap-1.2.1/./tests/./test78.sh: 4: ../examples/test21: not found
```

The contents of this test script are:

```
#!/bin/sh

# success
../examples/test21 ~~reverse /n mike > tmp.out 2>&1

if cmp -s tmp.out $srcdir/test78.out; then
	exit 0
else
	exit 1
fi
```

So... this is not making the out of tree builds happy. Setting the kwarg `test(workdir: ...)` to the build directory's tests/ directory should fix it I think.

**Comment by ManuelAtWork** [comment]
Just ammended my commit.

- Fixed path errors with unit tests.
- Replaced tabs with 2 spaces.

**Comment by jpakkane** [comment]
Done, thanks.

**Pull request merged**